homeassistant/components/search/__init__.py

```diff
@@ -0,0 +1,132 @@
+"""The Search integration."""
+from collections import defaultdict
+
+import voluptuous as vol
+
+from homeassistant.components import websocket_api
+from homeassistant.core import HomeAssistant, split_entity_id
+from homeassistant.helpers import device_registry, entity_registry
+
+DOMAIN = "search"
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Search component."""
+    websocket_api.async_register_command(hass, websocket_search)
+    return True
+
+
+@websocket_api.async_response
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "search",
+        vol.Required("item_type"): vol.In(
+            ("area", "device", "entity", "script", "scene", "automation")
+        ),
+        vol.Required("item_id"): str,
+    }
+)
+async def websocket_search(hass, connection, msg):
+    """Handle search."""
+    searcher = Searcher(
+        hass,
+        await device_registry.async_get_registry(hass),
+        await entity_registry.async_get_registry(hass),
+    )
+    connection.send_result(
+        msg["id"], await searcher.search(msg["item_type"], msg["item_id"])
+    )
+
+
+class Searcher:
+    """Find related things."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        device_reg: device_registry.DeviceRegistry,
+        entity_reg: entity_registry.EntityRegistry,
+    ):
+        """Search results."""
+        self.hass = hass
+        self._device_reg = device_reg
+        self._entity_reg = entity_reg
+        self.results = defaultdict(set)
+        self._to_resolve = set()
+
+    async def search(self, item_type, item_id):
+        """Find results."""
+        self._add_or_resolve(item_type, item_id)
+
+        while self._to_resolve:
+            search_type, search_id = self._to_resolve.pop()
+            await getattr(self, f"_resolve_{search_type}")(search_id)
+
+        # Remove entry into graph from search results.
+        self.results[item_type].remove(item_id)
+        # Clean up entity results with types that are represented as entities
```

Comment:
Are we cleaning this up because we haven't implemented those resolutions yet?

Reply:
We're cleaning this up because a referenced scene is referenced by automations/scripts as an entity, and so will end up in the entity results too.

Reply:
Ok. Maybe put it as:
Clean up entity_id items, from the general "entity" type result, that are also found in the specific entity domain type.